docs/basic/tests.md

```diff
@@ -6,24 +6,37 @@ Deal can automatically test your functions. First of all, your function has to b
 1. All exceptions that function can raise are specified in [@deal.raises](side-effects).
 1. All pre-conditions are specified with [@deal.pre](values).
 
-Then use `deal.cases` to get test cases for the function. Every case is a callable object that gets no arguments. Calling it will call the original function with suppressing allowed exceptions.
-
-## Example
-
 ```python
 @deal.raises(ZeroDivisionError)
 @deal.pre(lambda a, b: a >= 0 and b >= 0)
 def div(a: int, b: int) -> float:
     return a / b
+```
+
+Then you can use `deal.cases` to generate test cases for the function. This is a polimorphic object that can be used in many ways.
 
+Here is the shortest way to create a test:
+
+```python
+test_div = deal.cases(div)
+```
 
-for case in deal.cases(div):
+It is enough for [pytest](https://docs.pytest.org/en/latest/) to find and run the test. Or you can run it manually by just calling it: `test_div()`.
+
+However, it is not scalable. What if we want to use a pytest fixture? What if we need to prepare something before running the test case? Or what if we want to check additional conditions? So, let's make a proper test function:
+
+```python
+# type annotations below are optional
+@deal.cases(div)
+def test_div(case: deal.TestCase) -> None:
     case()
 ```
 
+In this example, when we (or pytest) call `test_div()`, deal will generate test cases (using [hypothesis](https://hypothesis.readthedocs.io/en/latest/)) and run the function body for every case. The test function itself decides when to execute the test case. Here `case` is an instance of [deal.TestCase](../details/api.html#deal.TestCase) class. This form works as expected with pytest fixtures if the test case is the first argument and has the name `case`.
+
 ## CLI
 
-There is a CLI command named `test`. It extracts `@deal.has()` (without arguments) wrapped functions and runs `deal.cases` powered tests for it.
+There is a CLI command named `test`. It extracts `deal.pure` and `@deal.has()` (without arguments) wrapped functions and runs `deal.cases` powered tests for it.
```

Comment:
It runs only pure functions to avoid undesired side-effects. `deal.pure` and `deal.has` indicate that the function is pure. However, when running the pure function, all contracts are checked (including `deal.raises`, of course).